massa-node/src/main.rs

```diff
@@ -182,8 +181,66 @@ async fn launch(
     )
 }
 
-// TODO: IDEA identify it unreachable code?
-async fn run(cfg: node_config::Config) {
+async fn stop(
+    bootstrap_manager: Option<BootstrapManager>,
+    consensus_manager: ConsensusManager,
+    consensus_event_receiver: ConsensusEventReceiver,
+    pool_manager: PoolManager,
+    protocol_manager: ProtocolManager,
+    network_manager: NetworkManager,
+    api_private_handle: StopHandle,
+    api_public_handle: StopHandle,
+) {
+    // stop bootstrap
+    if let Some(bootstrap_manager) = bootstrap_manager {
+        bootstrap_manager
+            .stop()
+            .await
+            .expect("bootstrap server shutdown failed")
+    }
+
+    // stop public API
+    api_public_handle.stop();
+
+    // stop private API
+    api_private_handle.stop();
+
+    // stop consensus controller
+    let protocol_event_receiver = consensus_manager
+        .stop(consensus_event_receiver)
+        .await
+        .expect("consensus shutdown failed");
+
+    // stop pool controller
+    let protocol_pool_event_receiver = pool_manager.stop().await.expect("pool shutdown failed");
+
+    // stop protocol controller
+    let network_event_receiver = protocol_manager
+        .stop(protocol_event_receiver, protocol_pool_event_receiver)
+        .await
+        .expect("protocol shutdown failed");
+
+    // stop network controller
+    network_manager
+        .stop(network_event_receiver)
+        .await
+        .expect("network shutdown failed");
+}
+
+#[tokio::main]
+async fn main() {
+    // setup logging
+    tracing_subscriber::fmt()
+        .with_max_level(match SETTINGS.logging.level {
+            4 => Level::TRACE,
+            3 => Level::DEBUG,
+            2 => Level::INFO,
+            1 => Level::WARN,
+            _ => Level::ERROR,
+        })
+        .init();
+
+    // run
```

Comment:
Why did you move the run part outside of its function ?

Reply:
because having it in a separate function call have no purpose here